src/Compilers/CSharp/Portable/Syntax/SyntaxNodeExtensions.cs

```diff
@@ -71,9 +71,11 @@ internal static bool CanHaveAssociatedLocalBinder(this SyntaxNode syntax)
                 kind == SyntaxKind.ArgumentList ||
                 kind == SyntaxKind.ArrowExpressionClause ||
                 (syntax is ExpressionSyntax && 
-                    // All these nodes are valid scope designators due to the pattern matching feature.
+                    // All these nodes are valid scope designators due to the pattern matching and out vars features.
                     ((syntax.Parent as LambdaExpressionSyntax)?.Body == syntax ||
                      (syntax.Parent as SwitchStatementSyntax)?.Expression == syntax ||
+                     (syntax.Parent as ForStatementSyntax)?.Condition == syntax ||
+                     (syntax.Parent as ForStatementSyntax)?.Incrementors.FirstOrDefault() == syntax ||
                      (syntax.Parent as CommonForEachStatementSyntax)?.Expression == syntax));
```

Comment:
Consider extracting a helper method for `(syntax is ExpressionSyntax && ...)`, for use here and in `SemanticModel`. The helper method could potentially be simplified (to use `Kind` checks and a single `Parent` call for instance).